stl/inc/charconv

```diff
@@ -1225,52 +1225,116 @@ _NODISCARD inline bool _Should_round_up(
+    // const bool _Tail_bits = !_Has_zero_tail || (_Value & _Extra_bits_mask) != 0;
 
-    return (_Value >> _Shift) + _Should_round_up(_Lsb_bit, _Round_bit, _Tail_bits);
-}
+    // return (_Value >> _Shift) + _Should_round_up(_Lsb_bit, _Round_bit, _Tail_bits);
 
-// Converts the floating-point value [sign] 0.mantissa * 2^exponent into the correct form for _FloatingType and
-// stores the result into the _Result object. The caller must ensure that the mantissa and exponent are correctly
-// computed such that either [1] the most significant bit of the mantissa is in the correct position for the
-// _FloatingType, or [2] the exponent has been correctly adjusted to account for the shift of the mantissa that
-// will be required.
+    // Example for optimized implementation: Let _Shift be 8.
+    // Bit index: ...[8]76543210
+    //    _Value: ...[L]RTTTTTTT
+    // By focusing on the bit at index _Shift, we can avoid unnecessary branching and shifting.
 
-// This function correctly handles range errors and stores a zero or infinity in the _Result object
-// on underflow and overflow errors, respectively. This function correctly forms denormal numbers when required.
+    // Bit index: ...[8]76543210
+    //  _Lsb_bit: ...[L]RTTTTTTT
+    const uint64_t _Lsb_bit = _Value;
 
-// If the provided mantissa has more bits of precision than can be stored in the _Result object, the mantissa is
-// rounded to the available precision. Thus, if possible, the caller should provide a mantissa with at least one
-// more bit of precision than is required, to ensure that the mantissa is correctly rounded.
-// (The caller should not round the mantissa before calling this function.)
+    //  Bit index: ...9[8]76543210
+    // _Round_bit: ...L[R]TTTTTTT0
+    const uint64_t _Round_bit = _Value << 1;
+
+    // We can detect (without branching) whether any of the trailing bits are set.
+    // Due to _Should_round below, this computation will be used if and only if R is 1, so we can assume that here.
+    //      Bit index: ...9[8]76543210
+    //     _Round_bit: ...L[1]TTTTTTT0
+    // _Has_tail_bits: ....[H]........
+
+    // If all of the trailing bits T are 0, and _Has_zero_tail is true,
+    // then `_Round_bit - static_cast<uint64_t>(_Has_zero_tail)` will produce 0 for H (due to R being 1).
+    // If any of the trailing bits T are 1, or _Has_zero_tail is false,
+    // then `_Round_bit - static_cast<uint64_t>(_Has_zero_tail)` will produce 1 for H (due to R being 1).
+    const uint64_t _Has_tail_bits = _Round_bit - static_cast<uint64_t>(_Has_zero_tail);
+
+    // Finally, we can use _Should_round_up() logic with bitwise-AND and bitwise-OR,
+    // selecting just the bit at index _Shift.
+    const uint64_t _Should_round = ((_Round_bit & (_Has_tail_bits | _Lsb_bit)) >> _Shift) & uint64_t{1};
+
+    // This rounding technique is dedicated to the memory of Peppermint. =^..^=
+    return (_Value >> _Shift) + _Should_round;

```

Comment:
Does this produce an observable performance difference? @cbezault noted that sometimes, branchless logic can be slower than branching. When I implemented this technique in hex `to_chars`, it was vastly preferable to walking through printed `char`s backwards and incrementing/carrying, but I never actually profiled the difference between the "reference implementation" of branching and this branchless technique.

Performance testing can be hugely time-consuming so it isn't absolutely necessary here - alternatively, what is the code size effect of branching vs. branchless? (You can inspect that by compiling with `/O2 /FAsc` and looking at the number of bytes of instructions emitted - we can help with that.) If this improves the observable performance of `from_chars` we should definitely use branchless. If a speed improvement can't be observed (likely, because `from_chars` is so expensive) but the codegen for branchless is smaller by even 1 byte, we should use it. However, if neither are improved, I think we should consider using the simpler-to-understand branching technique.